test/libraries/ProtocolFeeLibrary.t.sol

```diff
@@ -39,35 +36,43 @@ contract ProtocolFeeLibraryTest is Test {
-        assertGe(ProtocolFeeLibrary.calculateSwapFee(self, lpFee), lpFee);
+        fee = uint24(ProtocolFeeLibrary.MAX_PROTOCOL_FEE - 1) << 12 | ProtocolFeeLibrary.MAX_PROTOCOL_FEE - 1;
+        assertTrue(ProtocolFeeLibrary.validate(fee));
+
+        fee = uint24(0) << 12 | uint24(0);
+        assertTrue(ProtocolFeeLibrary.validate(fee));
     }
 
-    function test_fuzz_calculateSwapFeeNeverEqualsMax(uint24 self, uint24 lpFee) public pure {
-        lpFee = uint24(bound(lpFee, 0, LPFeeLibrary.MAX_LP_FEE - 1));
-        self = uint24(bound(self, 0, ProtocolFeeLibrary.MAX_PROTOCOL_FEE));
-        assertLt(ProtocolFeeLibrary.calculateSwapFee(self, lpFee), LPFeeLibrary.MAX_LP_FEE);
+    function test_fuzz_validate(uint24 fee) public pure {
+        if ((fee >> 12 > ProtocolFeeLibrary.MAX_PROTOCOL_FEE) || (fee % 4096 > ProtocolFeeLibrary.MAX_PROTOCOL_FEE)) {
+            assertFalse(ProtocolFeeLibrary.validate(fee));
+        } else {
+            assertTrue(ProtocolFeeLibrary.validate(fee));
+        }
     }
 
     function test_calculateSwapFee() public pure {
-        uint24 self = uint24(ProtocolFeeLibrary.MAX_PROTOCOL_FEE);
-        uint24 lpFee = LPFeeLibrary.MAX_LP_FEE;
-        assertEq(ProtocolFeeLibrary.calculateSwapFee(self, lpFee), LPFeeLibrary.MAX_LP_FEE);
-
-        lpFee = 3000;
-        assertEq(ProtocolFeeLibrary.calculateSwapFee(self, lpFee), 3997);
-
-        lpFee = 0;
-        assertEq(ProtocolFeeLibrary.calculateSwapFee(self, lpFee), ProtocolFeeLibrary.MAX_PROTOCOL_FEE);
-
-        self = 0;
-        assertEq(ProtocolFeeLibrary.calculateSwapFee(self, lpFee), 0);
+        assertEq(
+            ProtocolFeeLibrary.calculateSwapFee(uint24(ProtocolFeeLibrary.MAX_PROTOCOL_FEE), LPFeeLibrary.MAX_LP_FEE),
+            LPFeeLibrary.MAX_LP_FEE
+        );
+        assertEq(ProtocolFeeLibrary.calculateSwapFee(uint24(ProtocolFeeLibrary.MAX_PROTOCOL_FEE), 3000), 3997);
+        assertEq(
+            ProtocolFeeLibrary.calculateSwapFee(uint24(ProtocolFeeLibrary.MAX_PROTOCOL_FEE), 0),
+            ProtocolFeeLibrary.MAX_PROTOCOL_FEE
+        );
+        assertEq(ProtocolFeeLibrary.calculateSwapFee(0, 0), 0);
+        assertEq(ProtocolFeeLibrary.calculateSwapFee(0, 1000), 1000);
+    }
 
-        lpFee = 1000;
-        assertEq(ProtocolFeeLibrary.calculateSwapFee(self, lpFee), 1000);
+    function test_fuzz_calculateSwapFee(uint24 self, uint24 lpFee) public pure {
+        lpFee = uint24(bound(lpFee, 0, LPFeeLibrary.MAX_LP_FEE));
+        self = uint24(bound(self, 0, ProtocolFeeLibrary.MAX_PROTOCOL_FEE));
+        // if lp fee is not the max, the swap fee should never be the max since the protocol fee is taken off first and then the lp fee is taken from the remaining amount
+        if (lpFee < LPFeeLibrary.MAX_LP_FEE) {
+            assertLt(ProtocolFeeLibrary.calculateSwapFee(self, lpFee), LPFeeLibrary.MAX_LP_FEE);
+        }
+        assertGe(ProtocolFeeLibrary.calculateSwapFee(self, lpFee), lpFee);
```

Comment:
mm ideally we would have an `assertEq` line here too 😭 
given our comment on the function says
```
equivalent to protocolFee + lpFee(1_000_000 - protocolFee) / 1_000_000
```
can you fuzz that in the `eq`?